### PR TITLE
Make sure database files always exist when used

### LINKED ISSUE #5042: The app crashes after deleting and adding a project again
#### Problem description
The app crashes after deleting and adding a project again.

```
Caused by: android.database.sqlite.SQLiteReadOnlyDatabaseException: attempt to write a readonly database (code 1032 SQLITE_READONLY_DBMOVED)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:890)
        at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:756)
        at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:66)
        at android.database.sqlite.SQLiteDatabase.delete(SQLiteDatabase.java:1729)
        at org.odk.collect.android.database.forms.DatabaseFormsRepository.deleteForms(DatabaseFormsRepository.java:239)
        at org.odk.collect.android.database.forms.DatabaseFormsRepository.delete(DatabaseFormsRepository.java:164)
        at org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer.synchronizeAndReturnError(FormsDirDiskFormsSynchronizer.java:101)
        at org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer.synchronize(FormsDirDiskFormsSynchronizer.java:44)
        at org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher.fetchFormDetails(ServerFormsDetailsFetcher.kt:51)
        at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:63)
        at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:39)
```

#### Steps to reproduce the problem
1. Crete Demo project
2. Download one form
3. Delete the project
4. Create Demo project again
5. Try to download forms.

#### Other information 
The issue is similar to https://github.com/getodk/collect/issues/4870 but it looks like not the same.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Reverts https://github.com/getodk/collect/pull/5045
Closes #5042 

This pr adds changes implemented (and then reverted) in #5043 back.
I just got rid of one commit: https://github.com/getodk/collect/pull/5043/commits/8402b65b831312c15a55191d1e0c54a05405c594 which included closing databases and was a cause of failing tests.

I don't think the solution without closing databases is very bad because it's basically what we have on master and it doesn't seem to cause any big problems (on master we have a cache where we store database helpers and we never remove anything from that cache so if we create 100 projects and then remove 99 of them we will still keep all database helper object with open databases).
Of course it would be good to properly handle closing databases but it looks like it would require a lot of investigation. I'm also not a fan of removing caching at all. I think creating new objects for every operation might be something that will only bring other bugs.

### CODE REVIEW

**Comment by lognaturel** [comment]
I think this history is going to be tricky to follow. How about actually bringing back the individual commit(s) from the original PR?

**Comment by grzesiek2010** [comment]
> I think this history is going to be tricky to follow. How about actually bringing back the individual commit(s) from the original PR?

right, done.

**Comment by grzesiek2010** [comment]
There is something wrong with new tests. I'm not going to investigate it right away because if @seadowg is strongly against adding this back it wouldn't make sense.

File: collect_app/src/main/java/org/odk/collect/android/database/DatabaseConnection.kt [hunk 1]
```diff
@@ -26,29 +27,28 @@ open class DatabaseConnection(
...
-                name,
-                null,
-                databaseVersion,
-                migrator
-            )
+    private val dbHelper: SQLiteOpenHelper
+        get() {
+            val databasePath = path + File.separator + name
+            if (openHelpers.containsKey(databasePath) && !File(databasePath).exists()) {
+                openHelpers.remove(databasePath)
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/database/DatabaseConnection.kt]: Let's add a comment here (probably referencing the original PR) around why we're not including a call to `close` here.

**Review by seadowg — DISMISSED** [review]
I think we should go ahead with this. We will end up leaking `SQLiteOpenHelper` instances for projects that get deleted, but as we've all talked about, that's quite a rare occurrence, and we don't expect it to have any large consequences. @grzesiek2010 could you please file an issue around this so that we can have a think about how we might improve our connection pooling for later release.

File: collect_app/src/main/java/org/odk/collect/android/database/DatabaseConnection.kt (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/database/DatabaseConnection.kt]: Ok I added a short comment and the issue https://github.com/getodk/collect/issues/5078

**Review by seadowg — APPROVED** [review]

**Comment by grzesiek2010** [comment]
It was tested before so probably we can merge without testing right?

**Pull request merged**